src/core/components/HoverIntent/index.js

```diff
@@ -0,0 +1,76 @@
+import React, { PropTypes } from 'react';
+
+export default class HoverIntent extends React.Component {
+  static propTypes = {
+    sensitivity: PropTypes.number,
+    interval: PropTypes.number,
+    onHoverIntent: PropTypes.func.isRequired,
+    onHoverIntentEnd: PropTypes.func.isRequired,
+    children: PropTypes.element.isRequired,
+  }
+
+  static defaultProps = {
+    sensitivity: 5,
+    interval: 100,
+  }
+
+  componentWillUnmount() {
+    this.isHoverIntended = false;
+    this.clearHoverIntentDetection();
+  }
+
+  onMouseOver = (e) => {
+    e.persist();
+    const currentTarget = e.currentTarget;
+
+    const sq = (x) => x * x;
+    const distanceSq = (p1, p2) => sq(p1.x - p2.x) + sq(p1.y - p2.y);
+
+    const sensitivitySq = sq(this.props.sensitivity);
+
+    const initialPosition = { x: e.clientX, y: e.clientY };
+    let previousPosition = initialPosition;
+    this.currentMousePosition = initialPosition;
+
+    this.interval = setInterval(() => {
+      const currentPosition = this.currentMousePosition;
+      if (distanceSq(initialPosition, currentPosition) > sensitivitySq &&
+        distanceSq(previousPosition, currentPosition) < sensitivitySq) {
+        this.clearHoverIntentDetection();
+        this.isHoverIntended = true;
+
+        // construct a fake event cloned from e, but with e's initial currentTarget
+        // (currentTarget will change as the event bubbles up the DOM.)
+        this.props.onHoverIntent(Object.assign({}, e, { currentTarget }));
```

Comment:
e.persist() will remove it from React's event pool so that it won't be reused for separate events, but e.currentTarget is a bit of a special beast, because it changes for _this_ event as it bubbles up the DOM. If we don't do this, currentTarget ends up being null, since it has bubbled all the way up and then been cleared.